[HiCache]: Support dynamic loading backends for hicache

hzh0425 opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation

The original implementation in cache_controller.py had hardcoded backend creation logic.

This pr Introduced a StorageBackendFactory that provides:
- Dynamic Backend Loading: Backends can be loaded at runtime without code modification
- Plugin Architecture: External developers can create custom backends through configuration

To integrate your own custom storage backend with SGLang HiCache, follow these steps:

- Create a class inheriting from HiCacheStorage and implement all required abstract methods.
- Install Package ‘pip install -e .’
- Launch SGLang Server
```
python3 -m sglang.launch_server \
    --model-path your-model \
    --enable-hierarchical-cache \
    --hicache-storage-backend dynamic \
    --hicache-storage-backend-extra-config '{"backend_name":"custom_backend_name", "module_path": "your_module_path", "class_name": "YourHiCache Class Name"}'
```
Configuration Parameters:
--hicache-storage-backend: dynamic
--hicache-storage-backend-extra-config: JSON config with:
  - backend_name: Custom backend name
  - module_path: Python module path
  - class_name: Implementation class name


## Modifications

<!-- Detail the changes made in this pull request. -->

## Accuracy Tests

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->

## Benchmarking and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

hzh0425 committed:
[HiCache]: Support dynamic loading backends for hicache

Co-authored-by: Teng Ma <sima.mt@alibaba-inc.com>

xiezhq-hermann commented on python/sglang/srt/server_args.py:
```diff
@@ -1923,9 +1923,8 @@ def add_cli_args(parser: argparse.ArgumentParser):
         parser.add_argument(
             "--hicache-storage-backend",
             type=str,
-            choices=["file", "mooncake", "hf3fs", "nixl"],
```
I think it's probably fine to still keep the choices

hzh0425 replied:
If backend options are hardcoded here, it seems impossible to dynamically select the names of other backends?

xiezhq-hermann commented on python/sglang/srt/managers/cache_controller.py:
```diff
@@ -275,43 +275,15 @@ def __init__(
-                raise NotImplementedError(
-                    f"Unsupported storage backend: {storage_backend}"
+            try:
+                self.storage_backend = StorageBackendFactory.create_backend(
```
this is neat : )

hzh0425 committed:
Optimize code style

hzh0425 commented on python/sglang/srt/mem_cache/storage/backend_factory.py:
```diff
@@ -0,0 +1,260 @@
+
+            # Use original initialization logic for built-in backends
+            if registry_entry["is_builtin"]:
+                return cls._create_builtin_backend(
```
You are right; this is not needed here., fixed

hzh0425 committed:
Enhance hicache storage backend configuration

- Updated the `--hicache-storage-backend` argument to include "dynamic" as a choice and improved help text for clarity.

hzh0425 committed:
Merge branch 'main' into hicache/dyamic-loading

stmatengss committed:
Merge branch 'main' into hicache/dyamic-loading

hzh0425 commented:
@xiezhq-hermann  PTLA~

hzh0425 committed:
Merge remote-tracking branch 'origin/main' into hicache/dyamic-loading1

# Conflicts:
#	python/sglang/srt/managers/cache_controller.py
#	python/sglang/srt/server_args.py

hzh0425 committed:
Add aibrix to backend_factory

xiezhq-hermann committed:
Merge branch 'main' into hicache/dyamic-loading

hzh0425 committed:
Merge branch 'main' into hicache/dyamic-loading
